_posts/2020-12-04-release-1.12.0.md

```diff
@@ -0,0 +1,313 @@
+      <th>Description</th>
+      <th>Supported Connectors</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td><a href="https://ci.apache.org/projects/flink/flink-docs-release-1.12/dev/table/connectors/formats/avro-confluent.html">Avro Schema Registry</a></td>
+      <td>Read and write data serialized with the Confluent Schema Registry <a href="https://docs.confluent.io/platform/current/schema-registry/serdes-develop/serdes-avro.html">KafkaAvroSerializer</a>.</td>
+      <td>Kafka, Upsert Kafka</td>
+    </tr>
+    <tr>
+      <td><a href="https://ci.apache.org/projects/flink/flink-docs-release-1.12/dev/table/connectors/formats/debezium.html">Debezium Avro</a></td>
+        <td>Read and write Debezium records serialized with the Confluent Schema Registry KafkaAvroSerializer.</td>
+        <td>Kafka</td>
+    </tr>
+      <td><a href="https://ci.apache.org/projects/flink/flink-docs-release-1.12/dev/table/connectors/formats/raw.html">Raw</a></td>
+      <td>Read and write raw (byte-based) values as a single column.</td>
+      <td>
+        <p>Kafka, Upsert Kafka</p>
+        <p>Kinesis</p>
+        <p>FileSystem</p>
+      </td>
+    </tr>
+  </tbody>
+</table>
+
+**Multi-input Operator for Join Optimization ([FLINK-19621](https://issues.apache.org/jira/browse/FLINK-19621))**
+
+Shuffling is one of the most expensive operations in a Flink job. To eliminate unnecessary shuffles and improve the performance of batch and streaming Table API/SQL jobs, the default planner now leverages the N-ary stream operator introduced in the last release ([FLIP-92](https://cwiki.apache.org/confluence/display/FLINK/FLIP-92%3A+Add+N-Ary+Stream+Operator+in+Flink)) to implement the "chaining" of operators connected by forward edges and change these network shuffles to local function calls.
+
+**Type Inference for Table API UDAFs ([FLIP-65](https://cwiki.apache.org/confluence/display/FLINK/FLIP-65%3A+New+type+inference+for+Table+API+UDFs))**
+
+This release concluded the work started in Flink 1.9 on a [new data type system]({{ site.DOCS_BASE_URL }}flink-docs-release-1.12/dev/table/types.html#data-types) for the Table API, with the exposure of aggregate functions (UDAFs) to the new type system. From Flink 1.12, UDAFs behave similarly to scalar and table functions, and support all data types.
+
+<hr>
+
+### PyFlink: Python DataStream API
+
+To expand the usability of PyFlink, this release introduces a first version of the Python DataStream API ([FLIP-130](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=158866298)) with support for stateless operations (e.g. Map, FlatMap, Filter, KeyBy). To give the Python DataStream API a try, you can [install PyFlink]({{ site.DOCS_BASE_URL }}flink-docs-release-1.12/dev/python/installation.html#installation-of-pyflink) and check out [this tutorial]({{ site.DOCS_BASE_URL }}flink-docs-release-1.12/dev/python/datastream_tutorial.html) that guides you through building a simple streaming application.
```

Comment:
Could we add a simple example for the newly added Python DataStream API? e.g.:

To expand the usability of PyFlink, this release introduces a first version of the Python DataStream API ([FLIP-130](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=158866298)) with support for stateless operations (e.g. Map, FlatMap, Filter, KeyBy):

```python
from pyflink.common.typeinfo import Types
from pyflink.datastream import MapFunction, StreamExecutionEnvironment


class MyMapFunction(MapFunction):

    def map(self, value):
        return value + 1


env = StreamExecutionEnvironment.get_execution_environment()
data_stream = env.from_collection([1, 2, 3, 4, 5], type_info=Types.INT())
mapped_stream = data_stream.map(MyMapFunction(), output_type=Types.INT())
mapped_stream.print()
env.execute("datastream job")
```

To give the Python DataStream API a try, you can [install PyFlink]({{ site.DOCS_BASE_URL }}flink-docs-release-1.12/dev/python/installation.html#installation-of-pyflink) and check out [this tutorial]({{ site.DOCS_BASE_URL }}flink-docs-release-1.12/dev/python/datastream_tutorial.html) that guides you through building a simple streaming application.

Reply:
This makes sense, yes — I'll push it in. Thanks, @WeiZhong94!